hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java

```diff
@@ -434,7 +433,12 @@ private boolean initializeFromFilesystem(String initializationTime, List<Metadat
             fileGroupCountAndRecordsPair = initializeRecordIndexPartition();
             break;
           case FUNCTIONAL_INDEX:
-            fileGroupCountAndRecordsPair = initializeFunctionalIndexPartition();
+            Set<String> functionalIndexPartitionsToInit = getFunctionalIndexPartitionsToInit();
+            if (functionalIndexPartitionsToInit.isEmpty()) {
+              continue;
```

Comment:
Isn't line 392 already hanling the repetitive loading?

Reply:
No, i changed the logic slightly. `isMetadataPartitionAvailable` in line 392 only checks for func_index prefix in the metadata partition name. By that check, even an old functional index will pass for which we don't need to initialize. So, we get the uninitialized functional index partitions based on index definition in `getFunctionalIndexPartitionsToInit`.

Reply:
Can we unifiy the logic into one shot?

Reply:
We cannot do it currently without really getting rid of `MetadataPartitionType` enum. I think we should just have strings insetad of enum. If you agree, I can take up in a followup patch?

Reply:
I didn't get why the enum type affects the unification, can you elaborate a little bit?

Reply:
@danny0405 `MetadataPartitionType` has an inherent assumption that it maps directly to the physical partition (or rather partition path) for metadata table. While it is true for most of the MDT partitions, it is not the case with functional-index partitions (and the upcoming secondary index partition). `MetadataPartitionType` will only have a single entry for functional-index, `func_index_`, which acts as the prefix of the actual physical partition. The actual physical partitions is manifested as `func_index_<index-name>`. This is what the PR is trying to fix.

Reply:
Adding to what Vinay said, going forward we will have more indexes where index type and index name (mdt partition name) will differ such as secondary index. I think we should get rid of `MetadataPartitionType`.  This also enables removing the `MetadataRecordsGenerationParams` pojo which is deprecated.